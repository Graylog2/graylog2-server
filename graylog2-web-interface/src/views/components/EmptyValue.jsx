// @flow strict
import * as React from 'react';
import styled from 'styled-components';

const Container: React.ComponentType<{}> = styled.i`
  color: darkgray;
`;

const EmptyValue = () => <Container>&lt;Empty Value&gt;</Container>;

export default EmptyValue;
