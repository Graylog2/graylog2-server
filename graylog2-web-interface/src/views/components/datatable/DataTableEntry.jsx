// @flow strict
import React from 'react';
import * as Immutable from 'immutable';
import { flatten, get } from 'lodash';

import Value from 'views/components/Value';
import FieldType from 'views/logic/fieldtypes/FieldType';
import { AdditionalContext } from 'views/logic/ActionContext';
import type { ValuePath } from 'views/logic/valueactions/ValueActionHandler';
import Series, { parseSeries } from 'views/logic/aggregationbuilder/Series';
import type { FieldTypeMappingsList } from 'views/stores/FieldTypesStore';
import fieldTypeFor from 'views/logic/fieldtypes/FieldTypeFor';
import type { CurrentViewType } from '../CustomPropTypes';
import CustomHighlighting from '../messagelist/CustomHighlighting';
import DecoratedValue from '../messagelist/decoration/DecoratedValue';

type Props = {
  columnPivots: Array<string>,
  columnPivotValues: Array<Array<string>>,
  currentView: CurrentViewType,
  fields: Immutable.Set<string>,
  item: { [string]: * },
  series: Array<Series>,
  types: FieldTypeMappingsList,
  valuePath: ValuePath,
};

const _c = (field, value, path) => ({ field, value, path });

const _column = (field: string, value: *, selectedQuery: string, idx: number, type: FieldType, valuePath: ValuePath) => (
  <td key={`${selectedQuery}-${field}=${value}-${idx}`}>
    <AdditionalContext.Provider value={{ valuePath }}>
      <CustomHighlighting field={field} value={value}>
        {value !== null && value !== undefined ? <Value field={field} type={type} value={value} queryId={selectedQuery} render={DecoratedValue} /> : null}
      </CustomHighlighting>
    </AdditionalContext.Provider>
  </td>
);

const fullValuePathForField = (fieldName, valuePath) => {
  const currentSeries = parseSeries(fieldName);
  return currentSeries && currentSeries.field ? [...valuePath, { _exists_: currentSeries.field }] : valuePath;
};

const columnNameToField = (column, series = []) => {
  const currentSeries = series.find(s => s.effectiveName === column);
  return currentSeries ? currentSeries.function : column;
};

const DataTableEntry = ({ columnPivots, currentView, fields, series, columnPivotValues, valuePath, item, types }: Props) => {
  const classes = 'message-group';
  const { activeQuery } = currentView;

  const fieldColumns = fields.toSeq().toJS().map((fieldName, i) => _c(
    fieldName,
    item[fieldName],
    fullValuePathForField(fieldName, valuePath).slice(0, i + 1),
  ));
  const columnPivotFields = flatten(columnPivotValues.map((columnPivotValueKeys) => {
    const translatedPath = flatten(columnPivotValueKeys.map((value, idx) => [columnPivots[idx], value]));
    const [k, v]: Array<string> = translatedPath;
    const parentValuePath = [...valuePath, { [k]: v }];
    return series.map(({ effectiveName, function: fn }) => {
      const fullPath = [].concat(translatedPath, [effectiveName]);
      const value = get(item, fullPath);
      return _c(effectiveName, value, fullValuePathForField(fn, parentValuePath));
    });
  }));

  const columns = flatten([fieldColumns, columnPivotFields]);
  return (
    <tbody className={classes}>
      <tr className="fields-row">
        {columns.map(({ field, value, path }, idx) => _column(field, value, activeQuery, idx, fieldTypeFor(columnNameToField(field, series), types), path.slice()))}
      </tr>
    </tbody>
  );
};

export default DataTableEntry;
