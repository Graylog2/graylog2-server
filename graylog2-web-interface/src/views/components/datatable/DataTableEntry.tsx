/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import React from 'react';
import * as Immutable from 'immutable';
import { flatten, get } from 'lodash';
import styled, { css, DefaultTheme } from 'styled-components';

import Value from 'views/components/Value';
import FieldType from 'views/logic/fieldtypes/FieldType';
import { AdditionalContext } from 'views/logic/ActionContext';
import type { ValuePath } from 'views/logic/valueactions/ValueActionHandler';
import Series, { parseSeries } from 'views/logic/aggregationbuilder/Series';
import type { FieldTypeMappingsList } from 'views/stores/FieldTypesStore';
import fieldTypeFor from 'views/logic/fieldtypes/FieldTypeFor';

import type { CurrentViewType } from '../CustomPropTypes';
import CustomHighlighting from '../messagelist/CustomHighlighting';
import DecoratedValue from '../messagelist/decoration/DecoratedValue';

const StyledTd = styled.td(({ isNumeric, theme }: { isNumeric: boolean, theme: DefaultTheme }) => css`
  ${isNumeric ? `font-family: ${theme.fonts.family.monospace};` : ''}
  ${isNumeric ? 'text-align: right' : ''}
`);

type Field = {
  field: string,
  source: string,
};
type Props = {
  columnPivots: Array<string>,
  columnPivotValues: Array<Array<string>>,
  currentView: CurrentViewType,
  fields: Immutable.Set<Field>,
  item: { [key: string]: any },
  series: Array<Series>,
  types: FieldTypeMappingsList,
  valuePath: ValuePath,
};

const _c = (field, value, path, source) => ({ field, value, path, source });

const _column = (field: string, value: any, selectedQuery: string, idx: number, type: FieldType, valuePath: ValuePath, source: string | undefined | null) => (
  <StyledTd isNumeric={type.isNumeric()} key={`${selectedQuery}-${field}=${value}-${idx}`}>
    <AdditionalContext.Provider value={{ valuePath }}>
      <CustomHighlighting field={source ?? field} value={value}>
        {value !== null && value !== undefined ? <Value field={source ?? field} type={type} value={value} queryId={selectedQuery} render={DecoratedValue} /> : null}
      </CustomHighlighting>
    </AdditionalContext.Provider>
  </StyledTd>
);

const fullValuePathForField = (fieldName, valuePath) => {
  const currentSeries = parseSeries(fieldName);

  return currentSeries && currentSeries.field ? [...valuePath, { _exists_: currentSeries.field }] : valuePath;
};

const columnNameToField = (column, series = []) => {
  const currentSeries = series.find((s) => s.effectiveName === column);

  return currentSeries ? currentSeries.function : column;
};

const DataTableEntry = ({ columnPivots, currentView, fields, series, columnPivotValues, valuePath, item, types }: Props) => {
  const classes = 'message-group';
  const { activeQuery } = currentView;

  const fieldColumns = fields.toSeq().toJS().map(({ field: fieldName, source }, i) => _c(
    fieldName,
    item[fieldName],
    fullValuePathForField(fieldName, valuePath).slice(0, i + 1),
    source,
  ));
  const columnPivotFields = flatten(columnPivotValues.map((columnPivotValueKeys) => {
    const translatedPath = flatten(columnPivotValueKeys.map((value, idx) => [columnPivots[idx], value]));
    const [k, v]: Array<string> = translatedPath;
    const parentValuePath = [...valuePath, { [k]: v }];

    return series.map(({ effectiveName, function: fn }) => {
      const fullPath = [].concat(translatedPath, [effectiveName]);
      const value = get(item, fullPath);

      return _c(effectiveName, value, fullValuePathForField(fn, parentValuePath), fn);
    });
  }));

  const columns = flatten([fieldColumns, columnPivotFields]);

  return (
    <tbody className={classes}>
      <tr className="fields-row">
        {columns.map(({ field, value, path, source }, idx) => _column(field, value, activeQuery, idx, fieldTypeFor(columnNameToField(field, series), types), path.slice(), source))}
      </tr>
    </tbody>
  );
};

export default DataTableEntry;
