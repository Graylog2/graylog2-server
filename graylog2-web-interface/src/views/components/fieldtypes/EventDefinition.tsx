/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { create, keyResolver, windowScheduler } from '@yornaath/batshit';
import { useQuery } from '@tanstack/react-query';

import { EventsDefinitions } from '@graylog/server-api';

import Spinner from 'components/common/Spinner';

const fetchEventDefinition = create({
  fetcher: async (ids: Array<string>) => EventsDefinitions.getById({ event_definition_ids: ids }),
  resolver: keyResolver('id'),
  scheduler: windowScheduler(10),
});

const EventDefinition = ({ value }: { value: string }) => {
  const { data: eventDefinition, isLoading } = useQuery({
    queryKey: ['event-definitions', 'batched', value],

    queryFn: () => fetchEventDefinition.fetch(value),
  });

  if (isLoading) {
    return <Spinner />;
  }

  return eventDefinition?.title ? <span>{eventDefinition.title}</span> : <i>Missing Event Definition: {value}</i>;
};

export default EventDefinition;
