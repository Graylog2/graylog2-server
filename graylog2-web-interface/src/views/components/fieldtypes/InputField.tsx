/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useQuery } from '@tanstack/react-query';

import useInputs from 'hooks/useInputs';
import usePluginEntities from 'hooks/usePluginEntities';

type Props = {
  value: string,
}

const useForwarderInput = (inputId: string, enabled: boolean) => {
  const { fetchForwarderInput } = usePluginEntities('forwarder')[0] ?? {};
  const { data: forwarderInput, isError, isLoading } = useQuery(
    ['forwarder', 'input', inputId],
    () => fetchForwarderInput(inputId),
    { enabled: fetchForwarderInput && enabled },
  );

  return (isLoading || isError) ? undefined : forwarderInput;
};

const InputField = ({ value }: Props) => {
  const inputsMap = useInputs();
  const forwarderInput = useForwarderInput(value, inputsMap && !inputsMap[value]);

  const inputTitle = inputsMap[value]?.title ?? forwarderInput?.title ?? value;

  return <span title={value}>{inputTitle}</span>;
};

export default InputField;
