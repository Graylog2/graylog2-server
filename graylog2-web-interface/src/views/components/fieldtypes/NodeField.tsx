/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useQuery } from '@tanstack/react-query';

import useNodeSummaries from 'hooks/useNodeSummaries';
import usePluginEntities from 'hooks/usePluginEntities';

type Props = {
  value: string,
}

const useForwarderNode = (nodeId: string, enabled: boolean) => {
  const { fetchForwarderNode } = usePluginEntities('forwarder')[0] ?? {};
  const { data: forwarderNode, isError, isLoading } = useQuery(
    ['forwarder', 'node', nodeId],
    () => fetchForwarderNode(nodeId),
    { enabled: fetchForwarderNode && enabled },
  );

  return (isLoading || isError) ? undefined : forwarderNode;
};

const NodeField = ({ value }: Props) => {
  const nodes = useNodeSummaries();
  const node = nodes?.[value];
  const forwarderNode = useForwarderNode(value, nodes && !node);
  const nodeTitle = (node ? `${node.short_node_id} / ${node.hostname}` : forwarderNode?.title) ?? value;

  return <span title={value}>{nodeTitle}</span>;
};

export default NodeField;
