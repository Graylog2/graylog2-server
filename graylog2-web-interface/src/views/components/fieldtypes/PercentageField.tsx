/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useMemo } from 'react';
import numeral from 'numeral';
import styled from 'styled-components';

type Props = {
  value: number,
}

const NumberCell = styled.span`
  float: right;
`;

const PercentageField = ({ value }: Props) => {
  const formatted = useMemo(() => numeral(value).format('0.00%'), [value]);

  return <NumberCell title={String(value)}>{formatted}</NumberCell>;
};

export default PercentageField;
