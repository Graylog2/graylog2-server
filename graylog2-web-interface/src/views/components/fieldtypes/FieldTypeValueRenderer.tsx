/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import type { PluginExports } from 'graylog-web-plugin/plugin';

import Timestamp from 'components/common/Timestamp';

import InputField from './InputField';
import NodeField from './NodeField';
import StreamsField from './StreamsField';
import PercentageField from './PercentageField';
import EventDefinition from './EventDefinition';

const FieldTypeValueRenderer: PluginExports['fieldTypeValueRenderer'] = [
  {
    type: 'date',
    render: (value: string, field, render) => (
      <Timestamp dateTime={value} render={render} field={field} format="complete" />
    ),
  },
  {
    type: 'boolean',
    render: (value, field, Component) => <Component value={String(value)} field={field} />,
  },
  { type: 'input', render: (value) => <InputField value={String(value)} /> },
  { type: 'node', render: (value) => <NodeField value={String(value)} /> },
  {
    type: 'streams',
    render: (value: string | Array<string>) => <StreamsField value={value} />,
  },
  { type: 'percentage', render: (value: number) => <PercentageField value={value} /> },
  {
    type: 'event-definition-id',
    render: (value: string) => <EventDefinition value={value} />,
  },
];

export default FieldTypeValueRenderer;
