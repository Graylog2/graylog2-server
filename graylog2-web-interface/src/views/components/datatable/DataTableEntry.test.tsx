/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { List, OrderedSet } from 'immutable';
import { mount } from 'wrappedEnzyme';

import FieldTypeMapping from 'views/logic/fieldtypes/FieldTypeMapping';
import { FieldTypes } from 'views/logic/fieldtypes/FieldType';
import Series from 'views/logic/aggregationbuilder/Series';
import SeriesConfig from 'views/logic/aggregationbuilder/SeriesConfig';

import DataTableEntry from './DataTableEntry';

import EmptyValue from '../EmptyValue';

const f = (source: string, field: string = source): { field: string, source: string } => ({ field, source });
const createFields = (fields: Array<string>) => OrderedSet(fields.map((field) => f(field)));
const fields = createFields(['nf_dst_address', 'count()', 'max(timestamp)', 'card(timestamp)']);
const item = {
  nf_dst_address: '192.168.1.24',
  nf_proto_name: {
    TCP: { 'count()': 20, 'max(timestamp)': 1554106041841, 'card(timestamp)': 14 },
    UDP: { 'count()': 64, 'max(timestamp)': 1554106041841, 'card(timestamp)': 16 },
  },
  'count()': 84,
  'max(timestamp)': 1554106041841,
  'card(timestamp)': 20,
};

const columnPivots = ['nf_proto_name'];

const columnPivotValues = [['TCP'], ['UDP']];
const currentView = { activeQuery: '6ca0ea05-6fc1-4f46-9b22-20a5baab7b0d' };

const series = [
  Series.forFunction('count()'),
  Series.forFunction('max(timestamp)'),
  Series.forFunction('card(timestamp)'),
];

const valuePath = [{ nf_dst_address: '192.168.1.24' }];

const seriesWithName = (fn, name) => Series.forFunction(fn)
  .toBuilder()
  .config(SeriesConfig.empty()
    .toBuilder()
    .name(name)
    .build())
  .build();

jest.mock('views/hooks/useActiveQueryId', () => () => 'foobar');

describe('DataTableEntry', () => {
  const SUT = (props) => (
    <table>
      <tbody>
        <DataTableEntry columnPivots={columnPivots}
                        columnPivotValues={columnPivotValues}
                        currentView={currentView}
                        fields={fields}
                        item={item}
                        series={series}
                        types={List([])}
                        valuePath={valuePath}
                        {...props} />
      </tbody>
    </table>
  );

  it('does not fail without types', () => {
    const wrapper = mount((
      <SUT />
    ));

    expect(wrapper).not.toBeEmptyRender();
    expect(wrapper).toExist();
  });

  it('provides field types for fields and series', () => {
    const types = [
      FieldTypeMapping.create('timestamp', FieldTypes.DATE()),
      FieldTypeMapping.create('nf_dst_address', FieldTypes.STRING()),
    ];

    const wrapper = mount((
      <SUT types={List(types)} />
    ));

    const fieldTypeFor = (fieldName) => wrapper.find(`Value[field="${fieldName}"]`).first().props().type;

    expect(fieldTypeFor('nf_dst_address')).toEqual(FieldTypes.STRING());
    expect(fieldTypeFor('count()')).toEqual(FieldTypes.LONG());
    expect(fieldTypeFor('max(timestamp)')).toEqual(FieldTypes.DATE());
    expect(fieldTypeFor('card(timestamp)')).toEqual(FieldTypes.LONG());
  });

  it('provides valuePath in context for each value', () => {
    const wrapper = mount((
      <SUT />
    ));

    expect(wrapper.find('Provider')
      .map((p) => p.props().value))
      .toMatchSnapshot();
  });

  it('does not render `Empty Value` for deduplicated values', () => {
    const fieldsWithDeduplicatedValues = createFields(['nf_dst_address', 'nf_dst_port']);
    const itemWithDeduplicatedValues = {
      nf_dst_port: 443,
    };
    const wrapper = mount((
      <SUT fields={fieldsWithDeduplicatedValues}
           item={itemWithDeduplicatedValues} />
    ));

    expect(wrapper).not.toContainReact(<EmptyValue />);
  });

  describe('resolves field types', () => {
    const timestampTypeMapping = FieldTypeMapping.create('timestamp', FieldTypes.DATE());

    it('for non-renamed functions', () => {
      const wrapper = mount((
        <SUT types={List([timestampTypeMapping])} />
      ));
      const valueFields = wrapper.find('Value[field="max(timestamp)"]');

      valueFields.forEach((field) => expect(field).toHaveProp('type', timestampTypeMapping.type));
    });

    it('for simple row with renamed function', () => {
      const renamedSeries = [seriesWithName('max(timestamp)', 'Last Timestamp')];
      const itemWithRenamedSeries = {
        'Last Timestamp': 1554106041841,
      };
      const fieldsWithRenamedSeries = OrderedSet([f('max(timestamp)', 'Last Timestamp')]);

      const wrapper = mount((
        <SUT columnPivots={[]}
             columnPivotValues={[]}
             fields={fieldsWithRenamedSeries}
             item={itemWithRenamedSeries}
             series={renamedSeries}
             types={List([timestampTypeMapping])}
             valuePath={[]} />
      ));
      const valueFields = wrapper.find('Value[field="max(timestamp)"]');

      expect(valueFields).toHaveLength(1);

      valueFields.forEach((field) => expect(field).toHaveProp('type', timestampTypeMapping.type));
    });

    it('for renamed functions', () => {
      const renamedSeries = [
        Series.forFunction('count()'),
        seriesWithName('max(timestamp)', 'Last Timestamp'),
        Series.forFunction('card(timestamp)'),
      ];
      const itemWithRenamedSeries = {
        nf_dst_address: '192.168.1.24',
        nf_proto_name: {
          TCP: { 'count()': 20, 'Last Timestamp': 1554106041841, 'card(timestamp)': 14 },
          UDP: { 'count()': 64, 'Last Timestamp': 1554106041841, 'card(timestamp)': 16 },
        },
        'count()': 84,
        'Last Timestamp': 1554106041841,
        'card(timestamp)': 20,
      };
      const fieldsWithRenamedSeries = OrderedSet([
        f('nf_dst_address'),
        f('count()'),
        f('max(timestamp)', 'Last Timestamp'),
        f('card(timestamp)'),
      ]);

      const wrapper = mount((
        <SUT fields={fieldsWithRenamedSeries}
             item={itemWithRenamedSeries}
             series={renamedSeries}
             types={List([timestampTypeMapping])} />
      ));
      const valueFields = wrapper.find('Value[field="max(timestamp)"]');

      expect(valueFields).toHaveLength(3);

      valueFields.forEach((field) => expect(field).toHaveProp('type', timestampTypeMapping.type));
    });
  });
});
