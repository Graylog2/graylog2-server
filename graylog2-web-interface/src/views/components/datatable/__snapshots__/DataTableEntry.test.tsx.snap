// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableEntry provides valuePath in context for each value 1`] = `
[
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "TCP",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "TCP",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "TCP",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "UDP",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "UDP",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
  {
    "valuePath": [
      {
        "nf_dst_address": "192.168.1.24",
      },
      {
        "nf_proto_name": "UDP",
      },
      {
        "_exists_": "timestamp",
      },
    ],
  },
]
`;
